### PR TITLE
VocabBuilder: Fix and improve non-touch device experience

### LINKED ISSUE #12026: Vocabulary builder won't show focus on NT
* KOReader version: 2024.04.153
* Device: K4

#### Issue
vocabulary builder does not  show focus on currently selected items on non touch devices, i.e underline or black box on selected item. This makes it incredibly difficult to navigate and use.

#### Steps to reproduce
Open vocabulary builder on NT and try to use it.

### PR DESCRIPTION
**Pull request by weijiuqiao**
It seems due to the `FocusManager` setsDirty `self.show_parent or self` when it focuses on a target item, the target item does not show focus when the manager has a show_parent.

Fix #12026

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12030)

### CODE REVIEW

File: plugins/vocabbuilder.koplugin/main.lua [hunk 1]
```diff
@@ -1287,6 +1293,7 @@ function VocabularyBuilderWidget:init()
         self.key_events.Close = { { Device.input.group.Back } }
         self.key_events.NextPage = { { Device.input.group.PgFwd } }
         self.key_events.PrevPage = { { Device.input.group.PgBack } }
+        self.key_events.ShowMenu = { { { "Menu" } }}
```
Comment (hius07) [inline plugins/vocabbuilder.koplugin/main.lua:1296]: Looks like a redundant pair of brackets.

Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/main.lua]: Fixed.

**Review by Frenzie — APPROVED** [review]

**Comment by Commodore64user** [comment]
testing this, works fine now. Only one minor thing, for consistency, the menu button should both open and close the hamburger menu, right now it can only be opened by it but not closed (and yes I know you can do it with the back button).

**Comment by weijiuqiao** [comment]
> the menu button should both open and close the hamburger menu

It's done.

**Comment by Commodore64user** [comment]
that last commit is not strictly necessary (at least on kindle) as there is support for long press there already but I am not really bothered by it either, I was in fact going to say that pressing "Menu" inside that widget was crashing the programme but that technically solves that so... one more thing I just discovered is that the "search" widget does not support back button presses either, you can technically exit it by using focus and selecting cancel but for consistency again, best to map it to "Back" button keys.

File: plugins/vocabbuilder.koplugin/main.lua [hunk 2]
```diff
@@ -77,7 +77,7 @@ local MenuDialog = FocusManager:extend{
 function MenuDialog:init()
     self.layout = {}
     if Device:hasKeys() then
-        self.key_events.Close = { { Device.input.group.Back } }
+        self.key_events.Close = { { { "Back", "Menu" } } }
```
Comment (Commodore64user) [inline plugins/vocabbuilder.koplugin/main.lua:80]: @Frenzie bringing your attention here as this might affect other non-touch devices...

Comment (Frenzie) [reply plugins/vocabbuilder.koplugin/main.lua:80]: Yes, we would want to maintain the full Back group.

In ButtonDialog (which is used in most places) we did this:
```lua
        if Device:hasKeys() then
            local back_group = util.tableDeepCopy(Device.input.group.Back)
            if Device:hasFewKeys() then
                table.insert(back_group, "Left")
                self.key_events.Close = { { back_group } }
            else
                table.insert(back_group, "Menu")
                self.key_events.Close = { { back_group } }
            end
```

Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/main.lua]: Changed accordingly.

**Comment by Commodore64user** [comment]
`local util = require("util")` is missing

**Comment by weijiuqiao** [comment]
Oops. Thanks for helping testing.

File: plugins/vocabbuilder.koplugin/main.lua [hunk 3]
```diff
@@ -43,6 +43,7 @@ local TextBoxWidget = require("ui/widget/textboxwidget")
 local TitleBar = require("ui/widget/titlebar")
 local ToggleSwitch = require("ui/widget/toggleswitch")
 local UIManager = require("ui/uimanager")
+local util = require("util")
```
Comment (Frenzie) [inline plugins/vocabbuilder.koplugin/main.lua:46]: Could you move it under WidgetContainer above _, see https://github.com/koreader/koreader-base/wiki/Coding-style#order-of-requires

Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/main.lua]: Done.

**Pull request merged**